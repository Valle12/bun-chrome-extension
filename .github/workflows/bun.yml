name: Bun CI

on:
  pull_request:
    branches: ["master"]

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4
      - uses: oven-sh/setup-bun@v2

      - run: cd config && bun link
      - run: bun install --frozen-lockfile
      - run: cd ..
      - run: bun test
