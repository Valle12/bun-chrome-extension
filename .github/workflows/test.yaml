name: Test dev and bce packages

on:
  pull_request:
    branches: [ "master" ]

jobs:
  build:
    runs-on: ${{ matrix.os }}

    strategy:
      matrix:
        os: [ ubuntu-latest, windows-latest, macos-latest ]

    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Setup Bun
        uses: oven-sh/setup-bun@v2

      - name: Install and Test dev package
        run: cd dev && bun install --frozen-lockfile && bun test

      - name: Install and Test bce package
        run: bun install --frozen-lockfile && bun test
