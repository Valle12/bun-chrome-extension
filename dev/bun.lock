{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "bun-chrome-extension-dev",
      "devDependencies": {
        "@types/bun": "latest",
        "@types/chrome": "latest",
      },
      "peerDependencies": {
        "typescript": "^5.6.3",
      },
    },
  },
  "packages": {
    "@types/bun": ["@types/bun@1.2.1", "", { "dependencies": { "bun-types": "1.2.1" } }, "sha512-iiCeMAKMkft8EPQJxSbpVRD0DKqrh91w40zunNajce3nMNNFd/LnAquVisSZC+UpTMjDwtcdyzbWct08IvEqRA=="],

    "@types/chrome": ["@types/chrome@0.0.299", "", { "dependencies": { "@types/filesystem": "*", "@types/har-format": "*" } }, "sha512-hPoHmFWUaKw2s3wDjUtMsvZqeh9lKL7nFQx6yaAYt7RxBKv/V3rDUaA4BgUR145nnpF6hpNhHaxhtcCSFJ5b7w=="],

    "@types/filesystem": ["@types/filesystem@0.0.36", "", { "dependencies": { "@types/filewriter": "*" } }, "sha512-vPDXOZuannb9FZdxgHnqSwAG/jvdGM8Wq+6N4D/d80z+D4HWH+bItqsZaVRQykAn6WEVeEkLm2oQigyHtgb0RA=="],

    "@types/filewriter": ["@types/filewriter@0.0.33", "", {}, "sha512-xFU8ZXTw4gd358lb2jw25nxY9QAgqn2+bKKjKOYfNCzN4DKCFetK7sPtrlpg66Ywe3vWY9FNxprZawAh9wfJ3g=="],

    "@types/har-format": ["@types/har-format@1.2.16", "", {}, "sha512-fluxdy7ryD3MV6h8pTfTYpy/xQzCFC7m89nOH9y94cNqJ1mDIDPut7MnRHI3F6qRmh/cT2fUjG1MLdCNb4hE9A=="],

    "@types/node": ["@types/node@20.12.14", "", { "dependencies": { "undici-types": "~5.26.4" } }, "sha512-scnD59RpYD91xngrQQLGkE+6UrHUPzeKZWhhjBSa3HSkwjbQc38+q3RoIVEwxQGRw3M+j5hpNAM+lgV3cVormg=="],

    "@types/ws": ["@types/ws@8.5.12", "", { "dependencies": { "@types/node": "*" } }, "sha512-3tPRkv1EtkDpzlgyKyI8pGsGZAGPEaXeu0DOj5DI25Ja91bdAYddYHbADRYVrZMRbfW+1l5YwXVDKohDJNQxkQ=="],

    "bun-types": ["bun-types@1.2.1", "", { "dependencies": { "@types/node": "*", "@types/ws": "~8.5.10" } }, "sha512-p7bmXUWmrPWxhcbFVk7oUXM5jAGt94URaoa3qf4mz43MEhNAo/ot1urzBqctgvuq7y9YxkuN51u+/qm4BiIsHw=="],

    "typescript": ["typescript@5.6.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-hjcS1mhfuyi4WW8IWtjP7brDrG2cuDZukyrYrSauoXGNgx0S7zceP07adYkJycEr56BOUTNPzbInooiN3fn1qw=="],

    "undici-types": ["undici-types@5.26.5", "", {}, "sha512-JlCMO+ehdEIKqlFxk6IfVoAUVmgz7cU7zD/h9XZ0qzeosSHmUJVOzSQvvYSYWXkFXC+IfLKSIffhv0sVZup6pA=="],
  }
}
