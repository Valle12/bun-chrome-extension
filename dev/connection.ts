import type { WebSocketType } from "./types";

export class Connection {
  ws: WebSocket | undefined;
  interval = 2000;

  async connect() {
    if (this.ws) return;

    try {
      await fetch("http://localhost:3000", { mode: "no-cors" });
    } catch (e) {
      console.log("Server not running :(");
      return;
    }

    this.ws = <WebSocket>new WebSocket("ws://localhost:3000");
    console.log("Connected to server!");
    this.addListener();
  }

  addListener() {
    if (!this.ws) return;

    this.ws.onmessage = event => {
      const type = event.data as WebSocketType;
      if (type === "reload") chrome.runtime.reload();
    };

    this.ws.onclose = () => {
      console.log("Connection closed!");
      console.clear();
      this.ws = undefined;
      this.retryConnect();
    };
  }

  retryConnect() {
    this.connect();
    setInterval(() => {
      this.connect();
    }, this.interval);
  }
}
