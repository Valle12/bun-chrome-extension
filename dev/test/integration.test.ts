import { afterEach, beforeAll, describe, expect, test } from "bun:test";
import { watch } from "fs";
import { rm } from "fs/promises";
import { resolve } from "path";
import { Connection } from "./resources/bceIntegration/connection";

describe("bce integration", () => {
  const cwd = resolve(import.meta.dir, "resources/bceIntegration");

  beforeAll(async () => {
    await rm(resolve(cwd, "dist"), { recursive: true, force: true });
  });

  afterEach(async () => {
    await rm(resolve(cwd, "dist"), { recursive: true, force: true });
  });

  test("test if watching and reloading works as expected", async () => {
    const proc = Bun.spawn({
      cmd: ["bun", "./../../../bce.ts", "--dev"],
      cwd,
      env: { ...process.env, LOCAL: "true" },
      stdout: "ignore",
    });

    let timeout: Timer;
    let counter = 0;
    const watcher = watch(
      cwd,
      { recursive: true },
      async (_event, filename) => {
        if (!filename?.includes("manifest.json")) return;
        if (timeout) clearTimeout(timeout);
        timeout = setTimeout(async () => {
          counter++;
          new Connection().connect();

          const manifest = await Bun.file(
            resolve(cwd, "dist/manifest.json")
          ).text();
          expect(manifest).toMatchSnapshot();

          const solaceDist = await Bun.file(
            resolve(cwd, "dist/src/solace.js")
          ).text();
          expect(solaceDist).not.toContain("export");
          expect(solaceDist).toMatchSnapshot();

          if (counter === 1) {
            const solace = Bun.file(resolve(cwd, "src/solace.ts"));
            const content = await solace.text();
            await Bun.write(solace, content);
          } else if (counter === 2) {
            watcher.close();
            proc.kill("SIGINT");
          }
        }, 50);
      }
    );

    await proc.exited;
  }, 10000);
});
