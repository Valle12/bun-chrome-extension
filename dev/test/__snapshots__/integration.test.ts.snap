// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`bce integration test if watching and reloading works as expected 1`] = `
"{
  "manifest_version": 3,
  "name": "Test",
  "version": "0.0.1",
  "background": {
    "service_worker": "compose.js",
    "type": "module"
  },
  "content_scripts": [
    {
      "js": [
        "src/solace.js"
      ]
    }
  ]
}"
`;

exports[`bce integration test if watching and reloading works as expected 2`] = `
"class o{constructor(){console.log("Solace")}}new o;
"
`;

exports[`bce integration test if watching and reloading works as expected 3`] = `
"{
  "manifest_version": 3,
  "name": "Test",
  "version": "0.0.1",
  "background": {
    "service_worker": "compose.js",
    "type": "module"
  },
  "content_scripts": [
    {
      "js": [
        "src/solace.js"
      ]
    }
  ]
}"
`;

exports[`bce integration test if watching and reloading works as expected 4`] = `
"class o{constructor(){console.log("Solace")}}new o;
"
`;
